piccolo/testing/model_builder.py

```diff
@@ -159,29 +151,110 @@ def _randomize_attribute(cls, column: Column) -> t.Any:
+            Column class to randomize.
+
+        """
+        default_mapper = cls.__DEFAULT_MAPPER
+        if not default_mapper:  # execute once only
+            for typ, callable_ in RandomBuilder.get_mapper().items():
+                default_mapper[typ] = callable_
+
+        # order matters
+        reg = {
+            **default_mapper,
+            **cls._get_local_mapper(column),
+            **cls._get_other_mapper(column),
+        }
+
+        if column.value_type == list:
+            reg[list] = partial(
+                RandomBuilder.next_list,
+                reg[t.cast(Array, column).base_column.value_type],
+            )
+        return MappingProxyType(reg)
+
+    @classmethod
+    def _get_local_mapper(cls, column: Column) -> t.Dict[t.Type, t.Callable]:
+        """
+        This classmethod encapsulates the desired logic, utilizing information
+        from the column.
+
+        :param column:
+            Column class to randomize.
+        """
+        local_mapper: t.Dict[t.Type, t.Callable] = {}
+
+        precision, scale = column._meta.params.get("digits") or (4, 2)
+        local_mapper[Decimal] = partial(
+            RandomBuilder.next_decimal, precision, scale
+        )
+
+        tz_aware = getattr(column, "tz_aware", False)
+        local_mapper[datetime.datetime] = partial(
+            RandomBuilder.next_datetime, tz_aware
+        )
+
+        if _length := column._meta.params.get("length"):
+            local_mapper[str] = partial(RandomBuilder.next_str, _length)
+
+        return local_mapper
+
+    @classmethod
+    def _get_other_mapper(cls, column: Column) -> t.Dict[t.Type, t.Callable]:
+        """
+        This is a hook that allows users to register their own random type
+        callable. If the callable has a parameter named `column`, we assist
+        by injecting `column` using `partial`.
+
+        :param column:
+            Column class to randomize.
+
+        Examples::
+
+            # a callable not utilizing column information
+            ModelBuilder.register_random_type(str, lambda: "piccolo")
+
+            # a callable utilizing the column information
+            def next_str(column: Column) -> str:
+                length = column._meta.params.get("length", 5)
+                return "".join("a" for _ in range(length))
+            )
+            ModelBuilder.register_random_type(str, next_str)
+
+        """
+        other_mapper: t.Dict[t.Type, t.Callable] = {}
+        for typ, callable_ in cls.__OTHER_MAPPER.items():
+            sig = inspect.signature(callable_)
+            if sig.parameters.get("column"):
+                other_mapper[typ] = partial(callable_, column)
+            else:
+                other_mapper[typ] = callable_
+        return other_mapper
+
+    @classmethod
+    def register_type(cls, typ: t.Type, callable_: t.Callable) -> None:
+        cls.__OTHER_MAPPER[typ] = callable_
+
+    @classmethod
+    def unregister_type(cls, typ: t.Type) -> None:
```

Comment:
Being able to register custom types is cool, but I wonder about the main use cases.

You can specify defaults at the moment:

```python
await ModeBuilder.build(MyTable, defaults={MyTable.some_column: "foo"})
```

I'm not against being able to override how types are handled, but we just need to articulate to users when it's appropriate vs using `defaults`.

Reply:
Thank you for bringing up this question, it prompted me to reflect on the code and its implications.

The main difference between using `default` and `register_type` lies in their respective purposes:
* Utilizing `default` is suitable when users generally find our provided `random logic` satisfactory, but they require a hardcoded value for a specific column on a one-time basis.
* On the other hand, employing `register_type` is appropriate when users desire a custom implementation for a type, effectively overwriting our default `random logic` for that specific type.
Here are three distinct use cases:
  * (1) Types not provided by us: For instance, the type like `next_decimal` is not available in the current release version, but users can implement their own logic for the type and inject it into the `ModelBuilder`.
  * (2) User preference for specific logic: In scenarios where we introduce new features, such as the shiny `next_decimal` logic (returning `decimal.Decimal` if `column.value_type` is `decimal.Decimal`), users may prefer the previous implementation or have specific requirements. With `register_type`, they have the flexibility to override the default behavior.
  * (3) Unanticipated user cases: This aspect is particularly valuable for registration. For example, consider a user who initially builds a successful e-commerce platform in the UK using `Piccolo`. Later, they expand into Asia and encounter legal requirements necessitating the storage of customer names in local languages. If `ModelBuilder` does not support non-English characters, users can register their own implementations to address this issue.

A draft test for situations (1) and (2) might look like this:
```python
class TableWithDecimal(Table):
    numeric = Numeric()
    numeric_with_digits = Numeric(digits=(4, 2))
    decimal = Decimal()
    decimal_with_digits = Decimal(digits=(4, 2))

class TestModelBuilder(unittest.TestCase):
    ...
    def test_registry_overwritten1(self):
        table = ModelBuilder.build_sync(TableWithDecimal)
        for key, value in table.to_dict().items():
            if key != "id":
                self.assertIsInstance(value, decimal.Decimal)

        def fake_next_decimal(column: Column) -> float:
            """will return `float` instead of `decimal.Decimal`"""
            precision, scale = column._meta.params["digits"] or (4, 2)
            return RandomBuilder.next_float(
                maximum=10 ** (precision - scale), scale=scale
            )

        ModelBuilder.register_type(decimal.Decimal, fake_next_decimal)
        overwritten_table = ModelBuilder.build_sync(TableWithDecimal)
        for key, value in overwritten_table.to_dict().items():
            if key != "id":
                self.assertIsInstance(value, float)
```
A draft test for situations (3) might look like this:
```python
class TestModelBuilder(unittest.TestCase):
    ...
    def test_registry_overwritten2(self):
        choices = "一二三"  # Chinese characters

        def next_str(length: int = 3) -> str:
            # Chinese names often consist of three Chinese characters
            return "".join(random.choice(choices) for _ in range(length))

        ModelBuilder.register_type(str, next_str)
        manager1 = ModelBuilder.build_sync(Manager)
        self.assertTrue(all(char_ in choices for char_ in manager1.name))

        poster1 = ModelBuilder.build_sync(Poster)
        self.assertTrue(all(char_ in choices for char_ in poster1.content))

        ModelBuilder.unregister_type(str)
        manager2 = ModelBuilder.build_sync(Manager)
        self.assertTrue(all(char_ not in choices for char_ in manager2.name))

        poster2 = ModelBuilder.build_sync(Poster)
        self.assertTrue(all(char_ not in choices for char_ in poster2.content))
```
The scenario is as follows: `Manager1` is a locally hired individual, while `Manager2` is dispatched from the UK. Both are working on the `poster` using their native languages.

Finally, I realized I had overlooked the magic behavior of Python's name mangling rules. For instance:
```python
>>> class ModelBuilder:
...     __OTHER_MAPPER = {}
...
>>> ModelBuilder.__OTHER_MAPPER
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'ModelBuilder' has no attribute '__OTHER_MAPPER'
>>> ModelBuilder._ModelBuilder__OTHER_MAPPER
{}
```
As a result, the previous test code might be a bit off. I need to use the following code for the setup and teardown phases for each test:
```python
    def setUp(self) -> None:
        ModelBuilder._ModelBuilder__OTHER_MAPPER.clear()  # type: ignore

    def tearDown(self) -> None:
        ModelBuilder._ModelBuilder__OTHER_MAPPER.clear()  # type: ignore
```

Reply:
Thanks for explaining the rationale behind it - it makes sense.

If I was to completely redesign `ModelBuilder`, I probably wouldn't have class methods. Instead of:

```python
await ModelBuilder.build(MyTable)
```

I would have:

```python
await ModelBuilder(some_option=True).build(MyTable)
```

So we can configure ModelBuilder's behaviour easier. For registering types we could have:

```python
custom_builder = ModelBuilder(types={...})

await custom_builder.build(MyTable)
```

We could allow the types to be passed in via the `build` method instead:

```python
await ModelBuilder.build(MyTable, types={...})
```

If `register` and `unregister` work globally, there are pros and cons. The main pro is you only need to set it up once (e.g. in a session fixture of Pytest). But if you were to somehow run your tests in parallel, it might be problematic.

What do you think?

Reply:
Thank you for sharing your thoughts with me. Personally, I am inclined towards the instance method approach. However, implementing this change might break the current interface. I propose a three-stage transition plan:

1. First stage: Utilize the concept of `descriptors` to distinguish between calls from the `class` or `instance`. Initially, we move the current implementation to the `class` branch to maintain user experience. Simultaneously, we start implementing the new concept in the `instance` branch, issuing an `experimental warning`.

2. Second stage: If the new concept gains appreciation from users or developers, we add a `deprecated warning` to the `class` branch.

3. Third stage: Remove the `class` branch and clean up the code to ensure all methods are `instance methods` by the end.

During the first two stages, we'll keep the `class` branch unchanged and encourage users to try out the new syntax and the new features. If we reach the third stage, users who prefer the `class` branch might need to adjust their habits from using `await ModelBuilder.build(...)` to `await ModelBuilder().build()` or `ModelBuilder.build_sync(...)` to `ModelBuilder().build_sync(...)`.


The concept of `descriptors` is relatively straightforward, but it can sometimes feel too magical to grasp. I often need a refresher before coding if I haven't touched it for a long time. Fortunately, we don't need the complex `__get__` and `__set__` logic for the `data descriptor`. A simple `non-data descriptor` should suffice for our use case. With the help of this [post](https://stackoverflow.com/questions/28237955/same-name-for-classmethod-and-instancemethod),  I've drafted a concept code as follows:
```python
import asyncio
import inspect
import typing as t
from concurrent.futures import ThreadPoolExecutor


def run_sync(coroutine: t.Coroutine):
    try:
        # We try this first, as in most situations this will work.
        return asyncio.run(coroutine)
    except RuntimeError:
        # An event loop already exists.
        with ThreadPoolExecutor(max_workers=1) as executor:
            future = executor.submit(asyncio.run, coroutine)
            return future.result()


class dichotomy:
    def __init__(self, f):
        self.f = f

    def __get__(self, instance, owner):
        cls_or_inst = instance if instance is not None else owner
        if inspect.iscoroutine(self.f):
            async def newfunc(*args, **kwargs):
                return await self.f(cls_or_inst, *args, **kwargs)
        else:
            def newfunc(*args, **kwargs):
                return self.f(cls_or_inst, *args, **kwargs)
        return newfunc


class ModelBuilder:
    def __init__(self, *args, **kwargs):
        self._types = "..."  # Some information for instance method

    @dichotomy
    async def build(self_or_cls, *args, **kwargs):
        if inspect.isclass(self_or_cls):
            print("called as a class method from build")
            cls = self_or_cls
            await cls._build()
        else:
            print("called as an instance method from build")
            self = self_or_cls
            await self._build()

    @dichotomy
    def build_sync(self_or_cls, *args, **kwargs):
        return run_sync(self_or_cls.build())

    @dichotomy
    async def _build(self_or_cls, *args, **kwargs):
        if inspect.isclass(self_or_cls):
            print("called as a class method from _build", end="\n"*2)
            cls = self_or_cls  # noqa: F841
            # Current implementation remains here.
        else:
            print("called as an instance method from _build")
            self = self_or_cls
            # Some information can be retrieved.
            print(f'{self._types=}', end="\n"*2)
            # Our new logics


async def main():
    print('Async ModelBuilder.build: ')
    await ModelBuilder.build()

    print('Async ModelBuilder().build: ')
    await ModelBuilder().build()

    print('Sync ModelBuilder.build: ')
    ModelBuilder.build_sync()

    print('Sync ModelBuilder().build: ')
    ModelBuilder().build_sync()

if __name__ == '__main__':
    asyncio.run(main())
```
```
Async ModelBuilder.build: 
called as a class method from build
called as a class method from _build

Async ModelBuilder().build: 
called as an instance method from build
called as an instance method from _build
self._types='...'

Sync ModelBuilder.build: 
called as a class method from build
called as a class method from _build

Sync ModelBuilder().build: 
called as an instance method from build
called as an instance method from _build
self._types='...'
```
Finally, I agree that making `register` and `unregister` work globally could make it challenging to verify test results in parallel scenarios. I might lean towards using `instance methods` for the registering issue again.

These are just rough ideas that came to mind. I'm open to further discussions and refinements.

Reply:
Using descriptors is an interesting idea. I've used them sparingly before - as you say, they're very powerful, but can be confusing.

There's a lot of really good ideas in this PR, and I don't want to bog things down. I wonder if we could add this in a subsequent PR.

Reply:
Certainly! Here are some options to consider for closing this PR:

* Closing the PR without merging any changes.
* Keeping the current code as is, while implementing the `next_enum` functionality and updating related code.
* Utilizing the latest commit of this PR while removing the option for users to register custom types.
* Merging the PR with its latest commit.
* Considering any other suggestions or alternatives.

I'm open to any of these choices. @dantownsend , what are your thoughts on this matter?

Reply:
@jrycw  Sorry for the delay on this - I haven't forgotten about it, I'm just trying to finish off a couple of PRs.

I'm interested to know your thoughts on this:

https://github.com/piccolo-orm/piccolo/pull/978

If you think it's a good idea or not.

Reply:
@dantownsend no worries at all. It's great to see the project progressing on various fronts. I'll make an effort to review it and share any opinions or feedback I may have.